### ISSUE TITLE
The complexity of Computers should be increased

### ISSUE THREAD

**Issue by gamax92** [MightyPirates/OpenComputers#1043]
I believe that OpenComputers is too noob friendly and should be increased in complexity. The major help in complexity is making it so computers aren't one time builds, over time stuff will damage.

**Cases should be more complex:**
Instead of the cases being the tier, the case will remain constant. Sorta like the current Server Racks
A case base will be used to place boards, PSU's, HDDs, FDD, and Fans into.
A case panel will allow the computer to be controlled, I.E powered on, reset, and indicates the computer's status.
A case cover will hide the contents of the computer's
Items can then be placed into the case and configured via the case's connector.

A separate case is needed for the Servers. The server case is thinner than the normal case

**Boards should be more complex:**
Boards will be placed into the case and then when in the case, right clicked on to configure
Boards should also render their contents when in the case with no case cover
Boards can be shift right clicked to extract from the case.
Boards cannot be removed while components are attached.
Boards will also not be tier'd, and will have no tier concept, the board components will however remained tier'd.

**Board Assembly:**
Boards will be created in an Assembler, where a Circuit board is placed inside the Assembler, then components get placed on top, the “tier” of the board is determined by the number of components and type of components that get placed onto the board.
Board components get placed onto the Circuit board, and the soldering tool is used to secure them down. The soldering tool will also be used to connect components together
Provided you've placed necessary components to make a working board, you'll have a functional board.

**New Items:**
Hammer tool – Used to smash iron and gold ingots down into plates
Wire cutters – Breaks down gold plates into small strips to use in cards, also used for iron plates to create solder spool
Welder tool – Used to combine plates into shapes
Solder tool – Used to secure components onto boards and to connect components together, also used to repair broken traces
Desolder tool – Used to remove damaged components from the board
Power Supply Unit – Provides power to the board and other various components. This component is tier'd and higher tiers of the PSU will allow for more components to function.
Fans – Used to cool down the cpu and to cool down the computer

**Changed Items:**
T3 Graphics Card have an optional power connector, causing them to run as a T3.5 Graphics Card
HDD's are blobs of data, and OS's will need to have built in filesystems to be able to use them normally.
FDD's are also blobs of data.
Floppy Disks will now have tiers, with different capacities.
Graphics Cards will also have a graphics mode, which will allow higher resolutions at the trade off of no text, making it simpler for images to be created.
Internet cards should also be capable of doing UDP packets

**Board Components (New Items):**
CPU holder – Used to place the CPU into
Memory Controller – A chip that that can access memory. The Controller is tier'd and higher tiers can access more memory slots
Memory Slots – Used to place memory into
Card Controller - A chip than can serve multiple unlimited card slots
Card Slots – Used to place cards into. Card slots are tiered and backwards compatible with previous tiers, Newer cards can be placed into older slots but will only damage cards fast.
RTC – Gives the computer to ability to retrieve minecraft time, else the computer will always believe it is powered on at the beginning of time.
EEPROM holder – Used to place BIOS chips into
Hard Drive controller – A chip that can serve up to two HDD's
Hard Drive slot – Board connector for HDD's
Floppy Drive controller – A chip that can serve one FDD
Floppy Drive slot – Board connector for the FDD
Battery Holder – Holds a battery so the BIOS chip can retain memory.
Fan Controller – A chip that controls two fans.
Fan Slot – Board connector for a fan.
Power connector – Provides power to the board.

**Board Components are connected via the soldering tool, if the wrong items are connected to the wrong controller, they can be fried:**
Slots with items in them will destroy the item fast.
Individual components will be destroyed.
Solder lines from the CPU attaching to multiple controllers will destroy both controllers

**Damage over Time:**
Components should also damage themselves over time as they are used:
Hard Drives will remain reliable up to a point time where data will start to be randomly corrupted, and then after another time point, fail to work entirely. Dead HDD's will boot the computer up to a point, and then hang the computer whenever a Dead HDD is accessed. Attempting to use a Dead HDD will corrupt its contents exponentially fast. After a while they will shatter and make a loud grinding noise.
Floppy drives will also start to corrupt, with non fixable data, over time the entire FDD will become dead.
Memory modules will up to a point, stop working and cause the computer to refuse to boot
Batteries will lose their power over time, when at a very low point, the EEPROM will forget its information, and when at 0 the battery leaks onto the board and destroys traces
Traces on the board will also randomly break
Cards will damage over time and stop functioning after a time point, causing the computer to refuse to boot.
CPU's without a fan will heat up and be destroyed
Cases without a fan will heat up, speeding up the destruction of the computer's contents, and burning the player if they touch the case.
Cases will burn the player if attempting to configure them while powered on.
PSU's will damage over time, and if attempting to use too many components, will damage quicker. When the PSU dies, it will explode.
Overtime fans will drop in effectiveness and will cause components to heat up, destroying them faster. After a time point the fans will stop spinning entirely and act as if they wern't there.
The solder and desolder tools are wireless and while rechargable, will not charge up to full strength over time.

**Tweaking:**
Components should be able to be over clocked for faster speeds, while as a trade off, damaging components faster.
An overclocked cpu will allow for a slight boost in running time before getting the yeild error.
Overclocked graphics cards will allow for more operations per tick.
Overclocked components will also increase power draw, which may cause faster PSU death.
Floppy Disks can also be formatted to have more space. Since corrupted parts are not resizable, corruption will effect more data at one time.
</joke>

**Comment by Vexatos** [comment]
Actually, it's quite a reasonable request considering OC has ore processing now.

**Comment by elfifae** [comment]
You giant i386 fanatic. >:O Down with ATX! Viva la Amiga!

**Comment by gamax92** [comment]
Amiga Computers are also a possibility as an addon, custom cases/items/architectures are possible already. This is just to clean up and re factor the current machines.

**Comment by Kubuxu** [comment]
I see one problem with that:
MORE micro-crafting.

Sorry but micro-crafting IMHO does not make mod more newbie-friendly. Spent 10min throwing items around crafting table, Oh it is wrong item, looks like you will have to spend 10 more.

P.S. That is probably the biggest wall of text I have ever seen on OC's GitHub.

**Comment by Vexatos** [comment]
@Kubuxu Seriously? :suspect:

**Comment by gamax92** [comment]
Yes, exactly! the more micro-crafting time wasting that we can do will help to kill off the weak and let the strong survive in peace. It's like a rite of passage.

**Comment by KJA1582** [comment]
How about some PowerPC components? We need MOAR diversity! Or RISC and CISC CPU's so I can build my own calculator (CISC of course)!

**Comment by fnuecke** [comment]
If I could sticky this, I would.

**Comment by skyem123** [comment]
While this would be fun in some ways, it would not be fun in others.
One problem is that filesystems as "blobs of data" would make it harder for some people to write programs, as using an external text editor is much more convient, especially for programs meant to run at boot (init.lua).

**Comment by Vexatos** [comment]
@skyem123 You don't get it, do you? :dizzy:

**Comment by skyem123** [comment]
@Vexatos huh?

**Comment by asiekierka** [comment]
Everyone please leave. This would work well as an addon, like GregTech is an addon to IC2.

**Comment by Vexatos** [comment]
@gamax92 How about also adding the ability for small boards to be used as computers themselves? Kind of like a Raspberry Pi or an Arduino (i.e. single-board MCUs and single-board computers).

**Comment by sulljason** [comment]
Most of this sounds tedious. Adding pointless recipes that don't improve the depth of the mod, which to me is more about the engineering challenges. I would rather have to figure out how to overcome hardware limitations to do new, interesting, and useful things than use NEI to mash out more recipes for the sake of "realism". Only part I am in favour of is bit rot to give the RAID array more of a use (I have yet to run into disk space issues.).

**Comment by usafphoenix** [comment]
i also support the idea of micro-crafting computer parts and even having shit like the fans/ram/hard drives/cpus break over time or, rather, more-likely to break after a certain time of over-heating once the fans stop working. say, the fans can last upto x-number of days, and if they aren't replaced and the computer begins to overheat, the components NBT data takes damage...and after a certain threshold of damage, the computer fails to boot. 
Additionally, there can be a repair-mechanic added in to replace parts or use portions of parts to rebuild/fix it.

**Comment by Vexatos** [comment]
@fnuecke Better lock this before stuff happens.

**Comment by natedogith1** [comment]
I like the idea of file systems being blobs of data, though I can see why that'd be an undesirable implementation.
I don't know if fnuecke would add it, but I feel like a lot of people would be in favor of the graphics mode for graphics cards.
Internet cards really _should_ be able to send UDP packets.

**Comment by sulljason** [comment]
I think the limitation to TCP is to stop DOSing. Since you have to get a reply and actually establish a connection to send data over TCP, but I could be wrong.

**Comment by fnuecke** [comment]
@Vexatos nah, it's a great place to trap all these ideas.

Regarding UDP: misuse could be held in check by throttling the speed at which the respective send method could be called. The actual reason it's no in, is that I really don't see much of a point in it, since _receiving_ UDP packets will never be added (server sockets opened by user code? I'd rather not).
If you can show me a few _good_ use-cases, I might consider it.

For blob FSes, use tapes :P

**Comment by Vexatos** [comment]
Y'all go join the Glorious Tape Drive Masterrace! :sunglasses:

**Comment by SeniorFight** [comment]
100% Agree.
We're made of the same blood.

I want dis. @fnuecke i'll kill you if you close this.

**Comment by gamax92** [comment]
Woah there sir, no need to make death threats, joking or not.

**Comment by HiTECNOLOGYs** [comment]
This one really is exciting.

Sorry for being rude, I'm really not good at all this, but let's be straightforward. Micro-crafting, seriously? It's not 2009 anymore, we have many tools to automate that: from automatic crafting tables that do one recipe at a time to sophisticated systems that can even calculate required amount of resources and do necessary crafts in order to make more complex items in single click of mouse. Isn't that what progress is about, getting rid of unnecessary work while increasing the number of features your mechanisms support? Nobody makes you do all this crafting yourself. Nowadays it's hardly possible to have just one mod installed so is it really that hard to find a place for, say, Buildcraft if you're lazy? This reason for not adding some of listed things to OC as either an addon (which is preferable indeed until everything settles) or built-in sounds to me like a conservative attempt to humiliate all those who took part in making this mod better including OP and devs. People are not that foolish, they _know_ about this micro-crafting so called problems. Instead of complaining, be constructive or even go implement autocrafting in OC (in case you forgot, it's possible: drones can craft). Can we now move on to more important things?

**Comment by rashdanml** [comment]
The main issue with micro-crafting is striking a balance between both extremes (next to none on one end and GregTech-y on the other). OC is in a good place balance-wise, with some realistic elements. 

While adding more realism to the mod makes it more interesting, you would have to be careful not to go overboard. There are limits beyond which the realism interferes with the gameplay, making it less fun. Auto-crafting requires a fair bit of setup before it can be automated (AE2 for instance, requires you to craft all of the steps manually in order to encode the patterns - at least, in my experience). The more micro-crafting steps you add, the more tedious the setup becomes. Multiply that by all of the items that would require auto-crafting. The second issue with more steps is time taken to craft - recipes and patterns will need to be arranged in just the right way to be able to auto-craft components in parallel and assemble them. More steps = more time, generally. 

I'm in favour of adding more features in the same flavour as current crafting recipes in OC (most components only require about 3-4 crafting steps at most), but not going overboard with the realism (damage over time, etc).

**Comment by GitBranch** [comment]
It's a giant impressive wall of text! Anyways ... I think this should be OC2. The old one will be OC1 (which I prefer) and they both will receive updates. The programs are interchangeable tho

**Comment by Shuudoushi** [comment]
While I do think that tiered cases are a bit odd, I don't think adding more micro crafting to the mod is the right way to go. Almost all of what's listed in the OP should be put into an add-on, not the main mod.

You must keep something in mind here people, if you make it too hard or too tedious, people won't want to use the mod. Sure some people will love the piss out of it, but you still run the great risk of chasing off new users.

Oh, and blobing filesystems is just fucking stupid, have you ever tried making an init file from scratch in the OC editor? It's a gigantic pain in the ass. Like vex said, you want blobed fliesystems, go join the "tape master race", or w/e.

**Comment by n00b7** [comment]
GUI Pictures With Recipies:
![2015-06-15_08 43 49](https://cloud.githubusercontent.com/assets/11874683/8161443/e7f5f304-1344-11e5-95ad-0bcfd90ec777.png)
![2015-06-15_09 28 42](https://cloud.githubusercontent.com/assets/11874683/8161444/e7f60bb4-1344-11e5-815f-381a835b71d6.png)
![Uploading 2015-06-15_08.43.54.png…]()

**Comment by n00b7** [comment]
i have pictures for what i think the guis for computer cases should look like also recipies
![2015-06-15_09 28 42](https://cloud.githubusercontent.com/assets/11874683/8161467/0ffdd7e0-1345-11e5-991c-0698f630a9c6.png)
![2015-06-15_08 43 54](https://cloud.githubusercontent.com/assets/11874683/8161468/1000380a-1345-11e5-8273-79e950fa7200.png)
![2015-06-15_08 43 49](https://cloud.githubusercontent.com/assets/11874683/8161469/10027494-1345-11e5-965a-a93ce5821f14.png)

:

**Comment by LucaDev** [comment]
+1

**Comment by LucaDev** [comment]
I would like SSD's (same as HDD but silent and possibly a bit faster)

**Comment by MFernflower** [comment]
I like some of the ideas being discussed here.... Variable sized SSD's sound like a wonderful idea and so does a battery-backed RTC module. The idea of adding single board computers and modifiable electronics sounds marvelous in theory, but I haven't a clue how on earth it would be implemented. (Possibly see Victorious3/Integrated-Circuits?) As for the other more greg-tech style crafting - I would save that for a separate mod..........

**Comment by elfifae** [comment]
Hey, I'll let the recent posters in on a little tip

(It's a joke thread)

EDIT: TIL github markdown lacks spoiler tag support

**Comment by MFernflower** [comment]
^ this whole concept gamax92 wrote up was a joke??!??!???!?!

**Comment by elfifae** [comment]
Yep. Ironically, data blobs for disks DID end up getting implemented as unmanaged disks.

**Comment by jadenPete** [comment]
+1

**Comment by Shuudoushi** [comment]
@MFernflower, gamax really has nothing better to do 90% of the time other than come up with crazy shit like this.. Though, just like data blobs, some of the shit really is getting added in one way or another, though 99% of the really crazy shit is staying on the floor lol.

**Comment by MFernflower** [comment]
While this whole thing was a shitty legpull.  - Some of the less crazy stuff talked about (RTC module, SSD, Custom circuitry + SBC's, Etc) could be useful - provided that @fnuecke wants to/can find a way to implement it

**Comment by gjgfuj** [comment]
This thread was for discussing this sort of ideas, and it still is. Even
though op was a joke, its still a suggestion thread.

**Comment by MFernflower** [comment]
Im wondering if it's worth extracting the good ideas from the bullshit and
creating a new issue just with the good feature request ideas.

**Comment by ghost** [comment]
If this kind of component breaking does get implemented, make sure to either make the creative computers case not break its components, or make a tier -1 (creative) of everything that doesnt break

**Comment by Shuudoushi** [comment]
The component breaking thing will like never be added to OC itself while on default settings. If your activate hardmode, that might be another thing entirely, at least then it would make sense.

**Comment by ghost** [comment]
Why not separate Computer case and motherboard
Add more types of motherboard in which different types, features and roles
Heat is also make it better. ODD (Optical Disk Drives) + USB Flash Drives and RAID Controller (RAID 0, 1 and 10)

**Comment by baltazarhowardivanb** [comment]
Dear @fnuecke please implement this awesome idea on version 1.6.x.x (upcoming version) and the transformer class error

**Comment by baltazarhowardivanb** [comment]
@fnuecke @fnuecke @fnuecke @fnuecke @fnuecke @fnuecke @fnuecke @fnuecke @fnuecke @fnuecke @fnuecke @fnuecke @fnuecke @fnuecke @fnuecke @fnuecke @fnuecke @fnuecke @fnuecke @fnuecke @fnuecke @fnuecke @fnuecke @fnuecke @fnuecke @fnuecke @fnuecke @fnuecke @fnuecke @fnuecke @fnuecke @fnuecke @fnuecke @fnuecke @fnuecke @fnuecke @fnuecke @fnuecke @fnuecke @fnuecke @fnuecke @fnuecke @fnuecke @fnuecke @fnuecke

**Comment by baltazarhowardivanb** [comment]
@ChristoCoder @Shuudoushi @gjgfuj @MFernflower @jadenPete @lamialily @LucaDev @n00b7 @GitBranch @rashdanml @HiTECNOLOGYs @gamax92 @SeniorFight @Vexatos @sulljason @natedogith1 @usafphoenix @Kubuxu @KJA1582 @skyem123 @asiekierka

**Comment by LizzyTrickster** [comment]
Stop spamming the issue tracker, it's not going to make anyone implement it any faster

**Comment by baltazarhowardivanb** [comment]
@LizzyTrickster I DO NOT
![images](https://cloud.githubusercontent.com/assets/18028520/13985773/90c34a9a-f139-11e5-9912-25fa8cd207b2.jpg)

**Comment by asiekierka** [comment]
You just did.-------- Wiadomość oryginalna --------Temat: Re: [OpenComputers] The complexity of Computers should be increased (#1043)Od: baltazarhowardivanb Do: MightyPirates/OpenComputers DW: Adrian Siekierka @LizzyTrickster I DO NOT

**Comment by bauen1** [comment]
Der @fnuecke Please follow the steps specified in Github documentation on "How to block a User" https://help.github.com/articles/blocking-a-user/
I'am pretty sure that i just saved you from having to google that one :laughing:

**Comment by vifino** [comment]
Also see the harassment note below that.

**Comment by bauen1** [comment]
Yep, @baltazarhowardivanb is violating TOS
G.10 You must not upload, post, host, or transmit unsolicited email, SMSs, or "spam" messages.

**Comment by baltazarhowardivanb** [comment]
@bauen1 @bauen1 @bauen1 @bauen1 @bauen1 @bauen1 @bauen1 @bauen1 @bauen1 @bauen1 @bauen1 @bauen1 @bauen1 @bauen1 @bauen1 @bauen1 @bauen1 @bauen1 @bauen1 @bauen1 @bauen1 @bauen1 @bauen1 @bauen1 @bauen1 @bauen1 @bauen1 @bauen1 @bauen1 @bauen1 @bauen1 @bauen1 @bauen1 @bauen1 @bauen1 @bauen1 @bauen1 @bauen1 @bauen1 @bauen1 @bauen1

**Comment by bauen1** [comment]
Why do you think im offline?
Btw, your predictable and your violating Github TOS, gonna report you now

**Comment by LizzyTrickster** [comment]
I think he’s referring to fnuecke

 



---
This email has been checked for viruses by Avast antivirus software.
https://www.avast.com/antivirus

**Comment by bauen1** [comment]
Yeah, btw i found out that you can give -1 to posts :smile:

**Comment by fnuecke** [comment]
Already blocked him, hence the delayed response, apologies to everyone subscribed to this issue; apparently that doesn't apply to organizations owned by me, and I can't find how to block him from the organization. Reported and locking this for the time being.

**Comment by n00b7** [comment]
#1715 might increase complexity a little

**Comment by joeyyellow2005** [comment]
I **love** this and I think that it should be implemented! It makes total sense and it will make OC more fun!

**Comment by aaronduce** [comment]
YES YOU LEGEND. WE NEED THIS! Ps. also add 486 DX2s thnx

**Comment by sambrashaw** [comment]
I think this would be quite a clever addon so that people have the choice to go simpler or more advanced in the playstyle. Overall a very good idea.

**Comment by ghost** [comment]
Maybe the damage thing could be implemented with a button in the mod settings for , but I love the ideas listed under "Cases should be more complex" in the original post. As for micro-crafting, though, save for an addon.

**Comment by tomasbrod** [comment]
> The actual reason it's no in, is that I really don't see much of a point in it, since receiving UDP packets will never be added (server sockets opened by user code? I'd rather not).
If you can show me a few good use-cases, I might consider it.

The only way data can be sent over a network from OS (win/lin) is to open a socket, even for a TCP. Calling bind or listen should not be allowed thou.
Example UDP: reading time of a timeserver.

TCP is no better than UDP. Allowing TCP connections to be established does expose improperly configured services on the server local network or localhost to attacks. Some (broken) os services assume localhost to be secure and allow connections from localhost or local network to access privileged stuff. Or allowing attacker to use MC+OC server as proxy, to hide his identity, for another attack.

**Comment by skyem123** [comment]
It's worth noting that OC blocks any connection to local addresses by default

**Comment by Darinth** [comment]
Admittedly, a large portion of my issue with OC complexity is being fixed (at least in theory, I'm currently not playing MC and so I'm only really trying to stay more passively aware of the community for the mods I enjoy) by the wiki/documentation problems. When I moved from computercraft to opencomputers I enjoyed the increased complexity the came with, but the documentation with a bit outdated which made getting everything up and running initially a pain. If I felt more confident in the wiki staying up to date, I'd probably have no issue with some of these changes... but I also just don't see a particularly good reason. I think the mod currently strikes a good balance. There's complexity to utilize it, and is functionally it's balancing point. In order to get the benefits you have to be able to understand enough about computers to be able to put one together and write the programs that make it work.

**Comment by Sakata-MC** [comment]
Frankly I'm not a fan of component wear-out, adding un-needed tedium, and other such nonsense. This is a game. I'd totally get it as a plugin or something, but most of these suggestions would make me just not use the mod if implemented without the ability to turn it off.  Also, the data loss bits in the HDD stuff... that is why RL computers with important data use RAID.

Plus if data loss in a game gets to being an annoying part of the mix, I'd want a way to implement cold spares, backup systems with failure detection and restoration systems.  Like the real world that people tend to want to implement but only the idiotic parts of. And since material scarcity in minecraft is trivial to overcome, I could just make a self-replicating backup/storage system that slowly eats the world....

So yeah, please keep the mod away from the ideas of the masochists, and let the gamers have some fun with it :D
